src/server/frontend_wayland/wl_surface.cpp

```diff
@@ -210,11 +232,37 @@ void mf::WlSurface::frame(uint32_t callback)
 void mf::WlSurface::set_opaque_region(std::experimental::optional<wl_resource*> const& region)
 {
     (void)region;
+    // This isn't essential, but could enable optimizations
 }
 
 void mf::WlSurface::set_input_region(std::experimental::optional<wl_resource*> const& region)
 {
-    (void)region;
+    if (region)
+    {
+        // On Ubuntu 17.10 and later, the following two statements do the same thing.
+        // The former fails to compile on Ubuntu 16.04.
+
+        // pending.input_shape = WlRegion::from(region.value())->rectangle_vector();
+
+        pending.input_shape =
+            std::experimental::optional<std::experimental::optional<std::vector<geom::Rectangle>>>{
+                std::experimental::optional<std::vector<geom::Rectangle>>{
+                    WlRegion::from(region.value())->rectangle_vector()}};
```

Comment:
That's ugly. Does it really need to be spelled out in full?

Doesn't this work?

```
    pending.input_shape =
        decltype(pending.input_shape)::value_type{WlRegion::from(region.value())->rectangle_vector()};
```
(My wife's using the laptop with 16.04, so I can't test right now)

Reply:
Sorry, I don't have a 16.04 either (thus why I made it as verbose as possible to be sure it worked first try).